ktor-shared/ktor-websockets/jvm/test/WebSocketReaderTest.kt

```diff
@@ -11,21 +11,43 @@ import kotlinx.coroutines.*
 import kotlin.test.*
 
 class WebSocketReaderTest {
+    private suspend fun writeFrame(channel: ByteWriteChannel, frame: Frame) {
+        val serializer = Serializer()
+        serializer.enqueue(frame)
+
+        while (serializer.hasOutstandingBytes) {
+            channel.write {
+                serializer.serialize(it)
+            }
+        }
+    }
+
     @Test
     fun testFailFragmentedControl() {
-        val serializer = Serializer()
         val channel = ByteChannel(true)
         runBlocking {
             val reader = WebSocketReader(channel, coroutineContext, 1000)
             val frame = Frame.Ping(ByteArray(126))
-            serializer.enqueue(frame)
-
-            while (serializer.hasOutstandingBytes) {
-                channel.write {
-                    serializer.serialize(it)
-                }
+            writeFrame(channel, frame)
+            assertFailsWith<ProtocolViolationException> {
+                reader.incoming.receive()
             }
+        }
+    }
+
+    @Test
+    fun testInterruptedContinuation() {
+        val channel = ByteChannel(true)
+        runBlocking {
+            val reader = WebSocketReader(channel, coroutineContext, 1000)
+            writeFrame(channel, Frame.Text(false, "Foo".toByteArray()))
+            writeFrame(channel, Frame.Ping("Ping!".toByteArray()))
+            writeFrame(channel, Frame.Text(true, "Foo".toByteArray()))
+            assertIs<Frame.Text>(reader.incoming.receive())
+            assertIs<Frame.Ping>(reader.incoming.receive())
+
             assertFailsWith<ProtocolViolationException> {
+                // serializer will re-send the opcode
```

Comment:
Should we fix this and make sure that opcode is 0 for non final frames?
And can you please add a success path test for handling control frame in the middle of fragmented ones?

Reply:
> Should we fix this and make sure that opcode is 0 for non final frames?

That's not quite right. The opcode has to be 0 for frames _following_ non-final frames. This means that we'd need to add additional state tracking to the serializer (similar to the parser). I think the serializer currently doesn't handle non-fin frames correctly.

>  And can you please add a success path test for handling control frame in the middle of fragmented ones?

That's the same issue. Since the Serializer can't/doesn't handle fragmentation, there's currently not a great way to intersperse control frames in between fragmented frames. Most of the testing I've done with this has been with the Autobahn test suite instead.